keps/sig-scheduling/3521-pod-scheduling-readiness/README.md

```diff
@@ -538,10 +538,8 @@ https://storage.googleapis.com/k8s-triage/index.html
 We expect no non-infra related flakes in the last month as a GA graduation criteria.
 -->
 
-Create a test with the following sequences:
+An e2e test was created in Alpha with the following sequences:
```

Comment:
Another thing (it seems I missed that for Alpha - or it was changed in the meantime)

for Downgrade section vs the section from  Notes/Constraints/... above:
```
New field disabled in Alpha but not scheduler extension: In Alpha, the new Pod field is disabled by default. However, the scheduler's extension point is activated no matter the feature gate is enabled or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting different enqueue plugins and wire with custom fields or conditions.
```

That doesn't play well with downgrade. Because where we get is actually:
- I first upgraded to a release when the feature is enabled
- I created a pod with schedulingGate
- I downgrade the pod - the schedulingGate field remains set as expected
- but... on scheduler the feature is still enabled, and this pod will keep pending forever...

That seems wrong to me. Or was that properly solved on scheduler side?
That requires a more detailed discussion in the doc.

Reply:
> From what I know this at best will reach Alpha in 1.27, and we can't rely on Alpha in Beta feature.
What are your plans here?

Fine-grained control on k8s fields is a general topic. So I'd like to move it to future work to not block beta.

> That doesn't play well with downgrade. Because where we get is actually:
>
> - I first upgraded to a release when the feature is enabled
> - I created a pod with schedulingGate
> - I downgrade the pod - the schedulingGate field remains set as expected
> - but... on scheduler the feature is still enabled, and this pod will keep pending forever...

If we downgrade the cluster, scheduler has to be restarted, and during its initialization, the scheduler plugin (to back this feature) will [not be enabled](https://github.com/kubernetes/kubernetes/blob/d2ff210c205a2f929ffc7db61efcc90babdcf8a4/pkg/scheduler/apis/config/v1/default_plugins.go#L63-L65) at all. So if the downgrade is successful, no matter the previously-created pod carries schedulingGates or not, it will be bypassed.

Actually the section "New field disabled in Alpha but not scheduler extension" refers to another scenario: users don't opt-in to this feature (i.e., always leave .spec.schedulingGates empty), instead, they (usually advanced scheduler plugin developers) choose to apply custom fields (labels/annotations), and craft their own custom plugin to back those fields.

Reply:
> If we downgrade the cluster, scheduler has to be restarted, and during its initialization, the scheduler plugin (to back this feature) will not be enabled at all. So if the downgrade is successful, no matter the previously-created pod carries schedulingGates or not, it will be bypassed.

Ok - so that seems correct.
But then I don't really understand this part:
https://github.com/kubernetes/enhancements/pull/3739/files#diff-e8d9aae23eeca866dc6f8a2914e984dd8a74c22c05b7d6bb2800d88b047ce0b1R316-R319
```
In Alpha, the new Pod field is disabled
by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting
different enqueue plugins and wire with custom fields or conditions
```

What does the extension point do then? I though it's resposible for handling the pods with schedulingGates set. If not, then what that is?

> instead, they (usually advanced scheduler plugin developers) choose to apply custom fields (labels/annotations), and craft their own custom plugin to back those fields.

I'm sorry, I think I'm not following. Can you please clarify?

Reply:
sure, this section is more for plugin developers. In a high level, the entire feature contains the following parts:

- a. Pod's spec, feature gate, and other misc bits describing the field `schedulingGates`
- b. The `SchedulingGates` scheduler plugin
- c. The underlying scheduler framework, i.e., the new `PreEnqueue` extension point

So, what this section tries to say is: a plugin developer can use part c) _only_, and craft their own a' and b' to accomplish their own feature set.

Reply:
> What does the extension point do then? I though it's resposible for handling the pods with schedulingGates set. If not, then what that is?

The extension point is a lower-level core layer, which is agnostic to any pod field. As explained above, the SchedulingGates plugin aims to handle the pods with schedulingGates. When we disable this feature, it means the SchedulingGates plugin along with the upper-level API machinery bits are disabled, but the extension point (like an outlet in the wall) is still there.